src/main/java/grondag/frex/api/event/ConditionalEvent.java

```diff
@@ -0,0 +1,36 @@
+package grondag.frex.api.event;
+
+import java.util.function.Function;
+
+public abstract class ConditionalEvent<C, T> {
+    protected T invoker;
+
+    /**
+     * Returns the invoker instance.
+     *
+     * <p>An "invoker" is an object which hides multiple registered
+     * listeners of type T under one instance of type T, executing
+     * them and leaving early as necessary.
+     *
+     * @return The invoker instance.
+     */
+    public final T invoker() {
+        return invoker;
+    }
+
+    /**
+     * Register a listener to the event.
+     *
+     * @param condition Condition under which the listener should be fired.
+     * @param listener The desired listener.
+     */
+    public abstract void register(Function<C, Boolean> condition, T listener);
```

Comment:
Why not `Predicate<C>` here instead of `Function<C, Boolean>`.  Would avoid any boxing and is idiomatic.